retrofit/src/main/java/retrofit2/ResultCallAdapterFactory.kt

```diff
@@ -0,0 +1,88 @@
+package retrofit2
+
+import java.io.IOException
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+import okhttp3.Request
+import okio.Timeout
+
+class ResultCallAdapterFactory : CallAdapter.Factory() {
+  override fun get(
+    returnType: Type,
+    annotations: Array<Annotation>,
+    retrofit: Retrofit
+  ): CallAdapter<*, *>? {
+    if (getRawType(returnType) != Result::class.java) {
+      return null
+    }
+
+    check(returnType is ParameterizedType) {
+      "Result must have a generic type (e.g., Result<T>)"
+    }
+
+    val responseType = getParameterUpperBound(0, returnType)
+    return ResultCallAdapter<Any>(responseType)
+  }
+}
+
+class ResultCallAdapter<T>(
+  private val responseType: Type
+) : CallAdapter<T, Call<Result<T>>> {
+
+  override fun responseType(): Type {
+    return responseType
+  }
+
+  override fun adapt(call: Call<T>): Call<Result<out T>> {
+    return ResultCall(call)
+  }
+}
+
+class ResultCall<T>(private val delegate: Call<T>) : Call<Result<out T>> {
+
+  override fun enqueue(callback: Callback<Result<T>>) {
+    delegate.enqueue(object : Callback<T> {
+      override fun onResponse(call: Call<T>, response: Response<T>) {
+        val result = if (response.isSuccessful) {
+          Result.success(response.body()!!)
+        } else {
+          Result.failure(HttpException(response))
+        }
+        callback.onResponse(this@ResultCall, Response.success(result))
+      }
+
+      override fun onFailure(call: Call<T>, t: Throwable) {
+        callback.onResponse(this@ResultCall, Response.success(Result.failure(t)))
+      }
+    })
+  }
+
+  override fun execute(): Response<Result<T>> {
+    return try {
+      val response = delegate.execute()
+      val result = if (response.isSuccessful) {
+        Result.success(response.body()!!)
```

Comment:
Don't use !! - there is always a better way to fail gracefully into the Result.failure case if this is null.